craft_parts/plugins/gradle_plugin.py

```diff
@@ -0,0 +1,165 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2015-2023,2024,2025 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The gradle plugin."""
+
+import os
+from pathlib import Path
+from typing import Literal, cast
+from urllib.parse import urlparse
+
+from overrides import override
+from pydantic import model_validator
+from typing_extensions import Self
+
+from . import validator
+from .java_plugin import JavaPlugin
+from .properties import PluginProperties
+
+GRADLE_PLUGIN_VALIDATOR_FILE = Path(__file__).parent / "gradle_plugin_validator.py"
+
+
+class GradlePluginProperties(PluginProperties, frozen=True):
+    """The part properties used by the gradle plugin."""
+
+    plugin: Literal["gradle"] = "gradle"
+
+    gradle_init_script: str = ""
+    gradle_init_script_parameters: list[str] = []
+    gradle_parameters: list[str] = []
+    gradle_task: str = "build"
+
+    # part properties required by the plugin
+    source: str  # pyright: ignore[reportGeneralTypeIssues]
+
+    @model_validator(mode="after")
+    def gradle_task_defined(self) -> Self:
+        """Gradle task must be defined."""
+        if not self.gradle_task:
+            raise ValueError("Gradle task must be defined")
+        return self
+
+
+class GradlePluginEnvironmentValidator(validator.PluginEnvironmentValidator):
+    """Check the execution environment for the gradle plugin."""
+
+    @override
+    def validate_environment(
+        self, *, part_dependencies: list[str] | None = None
+    ) -> None:
+        """Ensure the environment contains dependencies needed by the plugin.
+
+        :param part_dependencies: A list of the parts this part depends on.
+        :raises PluginEnvironmentValidationError: If gradle is invalid
+          and there are no parts named gradle-deps.
+        """
+        # We can't check for gradle executable here because if the project uses gradlew, we don't
+        # necessarily need gradle to be installed on the system. We can't check if the project uses
+        # gradlew since the project would not be pulled yet.
```

Comment:
So do we need the function at all? Or is it a required override?

Reply:
We don't need it but I thought it better be reasoned why it is not implemented as a comment :)

Reply:
Perhaps instead of a noop function that comment can go into the class docstring?